spring-kafka/src/main/java/org/springframework/kafka/support/converter/BatchMessagingMessageConverter.java

```diff
@@ -275,13 +288,22 @@ protected Object extractAndConvertValue(ConsumerRecord<?, ?> record, Type type)
 	 * @param type the type - must be a {@link ParameterizedType} with a single generic
 	 * type parameter.
 	 * @param conversionFailures Conversion failures.
-	 * @return the converted payload.
+	 * @return the converted payload, potentially further processed by a {@link SmartMessageConverter}.
 	 */
 	protected @Nullable Object convert(ConsumerRecord<?, ?> record, Type type, List<ConversionException> conversionFailures) {
 		try {
 			if (this.recordConverter != null) {
 				Object payload = this.recordConverter
 						.toMessage(record, null, null, ((ParameterizedType) type).getActualTypeArguments()[0]).getPayload();
+				if (this.messagingConverter != null && payload != null && !payload.equals(KafkaNull.INSTANCE)) {
+					Class<?> clazz = ((ParameterizedType) type).getActualTypeArguments()[0] instanceof Class ?
+							(Class<?>) ((ParameterizedType) type).getActualTypeArguments()[0] : Object.class;
+					Message<?> message = MessageBuilder.withPayload(payload).build();
+					Object convertedPayload = this.messagingConverter.fromMessage(message, clazz, ((ParameterizedType) type).getActualTypeArguments()[0]);
```

Comment:
You use `((ParameterizedType) type).getActualTypeArguments()[0]` four times.
Not efficient from the call stack perspective.
Please, consider to extract it into the local variable.
This way such a call would happen only once.